File: data-processing-system/java/Worker.java
public final class Worker implements Runnable {
    private final String workerName;
    private final SharedTaskQueue taskQueue;
    private final ResultStore resultStore;

    public Worker(String workerName, SharedTaskQueue taskQueue, ResultStore resultStore) {
        this.workerName = workerName;
        this.taskQueue = taskQueue;
        this.resultStore = resultStore;
    }

    @Override
    public void run() {
        System.out.println("[START] " + workerName + " started");

        while (true) {
            Task task;
            try {
                task = taskQueue.getTask();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.out.println("[ERROR] " + workerName + " interrupted while waiting for a task: " + e.getMessage());
                break;
            }

            if (task.isPoisonPill()) {
                System.out.println("[EXIT] " + workerName + " received shutdown signal");
                break;
            }

            System.out.println("[PROCESSING] " + workerName + " picked Task " + task.getTaskId() + " (payload=" + task.getPayload() + ")");

            try {
                ProcessedResult result = processTask(task);
                resultStore.addResult(result);
                if (result.isSuccess()) {
                    System.out.println("[COMPLETED] " + workerName + " finished Task " + task.getTaskId() + " -> result=" + result.getResultValue());
                } else {
                    System.out.println("[ERROR] " + workerName + " failed Task " + task.getTaskId() + " (payload=" + task.getPayload() + "): " + result.getMessage());
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                String message = "interrupted during processing";
                resultStore.addResult(new ProcessedResult(task.getTaskId(), task.getPayload(), workerName, 0, false, message));
                System.out.println("[ERROR] " + workerName + " " + message + " for Task " + task.getTaskId());
            } catch (IllegalArgumentException e) {
                resultStore.addResult(new ProcessedResult(task.getTaskId(), task.getPayload(), workerName, 0, false, e.getMessage()));
                System.out.println("[ERROR] " + workerName + " failed Task " + task.getTaskId() + " (payload=" + task.getPayload() + "): " + e.getMessage());
            }
        }

        System.out.println("[STOP] " + workerName + " exited cleanly");
    }

    private ProcessedResult processTask(Task task) throws InterruptedException {
        /*
         * The sleep is a simple simulation of real work. It also makes the
         * concurrency visible when the console output is captured.
         */
        Thread.sleep(150L + (task.getTaskId() % 3) * 75L);

        if (task.getPayload() < 0) {
            throw new IllegalArgumentException("negative payload is invalid");
        }

        int resultValue = task.getPayload() * task.getPayload();
        return new ProcessedResult(
            task.getTaskId(),
            task.getPayload(),
            workerName,
            resultValue,
            true,
            "Processed successfully"
        );
    }
}

File: data-processing-system/go/worker.go
package main

import (
	"fmt"
	"log"
	"time"
	"sync"
)

// worker runs as a goroutine. It reads from the task channel, processes each
// task, and sends a result to the results channel. The wait group lets main
// know when all workers have exited safely.
func worker(name string, tasks <-chan Task, results chan<- ProcessedResult, wg *sync.WaitGroup) {
	defer wg.Done()

	log.Printf("[START] %s started", name)

	for task := range tasks {
		log.Printf("[PROCESSING] %s picked Task %d (payload=%d)", name, task.ID, task.Payload)

		result := processTask(name, task)
		results <- result

		if result.Success {
			log.Printf("[COMPLETED] %s finished Task %d -> result=%d", name, task.ID, result.ResultValue)
		} else {
			log.Printf("[ERROR] %s failed Task %d (payload=%d): %s", name, task.ID, task.Payload, result.Message)
		}
	}

	log.Printf("[EXIT] %s exited cleanly", name)
}

func processTask(workerName string, task Task) ProcessedResult {
	// The sleep is a simple simulation of real work and helps make the
	// concurrent worker behavior visible in the console.
	time.Sleep(150*time.Millisecond + time.Duration(task.ID%3)*75*time.Millisecond)

	if task.Payload < 0 {
		return ProcessedResult{
			TaskID:      task.ID,
			Payload:     task.Payload,
			WorkerName:  workerName,
			ResultValue: 0,
			Success:     false,
			Message:     "negative payload is invalid",
		}
	}

	resultValue := task.Payload * task.Payload
	return ProcessedResult{
		TaskID:      task.ID,
		Payload:     task.Payload,
		WorkerName:  workerName,
		ResultValue: resultValue,
		Success:     true,
		Message:     "Processed successfully",
	}
}

func logTaskSummary(results []ProcessedResult) {
	fmt.Println()
	fmt.Println("========================================")
	fmt.Println("Final Results Summary")
	fmt.Println("========================================")
	for _, result := range results {
		fmt.Println(result.String())
	}
}
